### PR TITLE
Move QPY version regex construction to import time

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

As pointed out by @nkanazawa1989 in #8232 the regex construction added
to the QPY interface functions in #8200 can be a significant portion of
the overall function time especially for smaller inputs. Especially as
we're building it on every call interface function call. To ameliorate
that cost this commit compiles the version regex a single time at the
module level. This adds the overhead to import but it will only be done
once which should be a net improvement in performance.

### Details and comments

### CODE REVIEW

**Review by jakelishman — DISMISSED** [review]
Looks good to me - my quick search showed these two instances as the only uses of the regex as well (other than the test suite).

File: qiskit/qpy/interface.py [hunk 1]
```diff
@@ -69,6 +69,7 @@
 """
     + "$"
 )
+VERSION_PATTERN_REGEX = re.compile(VERSION_PATTERN)
```
Comment (jakelishman) [inline qiskit/qpy/interface.py:72]: Oh, the flags probably need moving out of the `search` calls and into the `compile`.

Comment (mtreinish) [reply qiskit/qpy/interface.py:72]: Oh good call, completely missed that. Let me fix that right now

Comment (mtreinish) [reply qiskit/qpy/interface.py]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/8235/commits/2760a38968b2ded1bd4cc2bf25665ec1e0a337df

**Pull request merged**